In [28]:
# Imports and pip installations (if needed)
import pandas as pd
import numpy as np
from scipy import stats

# Part 1: Load the dataset

In [29]:
# Load the given datasets

categorical = pd.read_csv("data/chronic_kidney_disease_categorical.csv")
numerical = pd.read_csv("data/chronic_kidney_disease_numerical.csv")
# Print the data

print(categorical)

print(numerical)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes   no     c

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: Each row of the data represents a set of readings and data about a patient. The features are various health indicators and other pieces of data about the patient that might be relevant to their kidney health, and they are labeled with whether they have chronic kidney disease or not. The features are split into two CSV tables based on whether they represent categorical or numerical features. The categorical features include whether the patient is diabetic (dm), whether pus cell clumps are present (pcc), and whether the patient is anemic (ane). THe numerical columns include the patient's age (age), blood pressure (bp), and hemoglobin levels (hemo).

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [30]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
if (numerical.duplicated().sum() != 0):
    print(numerical.duplicated().sum())


# Drop these duplicate rows
numerical.drop_duplicates()

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
if (categorical.duplicated().sum != 0):
    print(categorical.duplicated().sum())
    categorical.drop_duplicates()


30
30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: Not all the unique ids are present in both datasets, which I know because the inner merge has less data than the outer merge. In the combined table, rows missing from one of the datasets has NaN for all the features that are from the dataset that id isn't present in.

In [31]:
# Merge the two given numerical and categorical datasets based on their unique_ID.

joined = pd.merge(categorical, numerical, how="outer", on="unique_id")
#Print the combined dataset
print(joined)

     unique_id   al   su     rbc        pc         pcc          ba  htn   dm  \
0       102677  3.0  0.0  normal  abnormal  notpresent  notpresent  yes   no   
1       103570  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
2       104675  2.0  1.0  normal    normal     present  notpresent   no  yes   
3       105923  1.0  0.0     NaN    normal  notpresent  notpresent  yes  yes   
4       107297  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
..         ...  ...  ...     ...       ...         ...         ...  ...  ...   
487     995177  4.0  3.0  normal  abnormal     present     present  yes  yes   
488     996076  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
489     996412  NaN  NaN     NaN       NaN  notpresent  notpresent   no   no   
490     997629  0.0  0.0  normal    normal         NaN         NaN   no   no   
491     997876  0.0  0.0     NaN    normal  notpresent  notpresent   no   no   

     cad  ...    bp    bgr     bu    sc

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [32]:
# Calculate the percentage of rows that contain atleast one missing value
countmissing = joined.isna().any(axis=1).sum()
counttotal = joined.shape[0]
percentage = countmissing/counttotal * 100
# Print %
print("Percent missing at least one value: ")
print(percentage)

# Drop these rows from the dataset
joined.dropna(inplace = True)
# Print the Dataset
print(joined)

Percent missing at least one value: 
59.756097560975604
     unique_id   al   su     rbc        pc         pcc          ba  htn   dm  \
0       102677  3.0  0.0  normal  abnormal  notpresent  notpresent  yes   no   
6       109053  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
7       114717  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
8       118191  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
9       118191  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
..         ...  ...  ...     ...       ...         ...         ...  ...  ...   
482     980291  3.0  0.0  normal  abnormal  notpresent  notpresent  yes  yes   
483     991031  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
486     994377  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
487     995177  4.0  3.0  normal  abnormal     present     present  yes  yes   
488     996076  0.0  0.0  normal    normal  notpresent  notprese

## Part 3.4: Sort the dataset according to the Labels

In [33]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
joined.sort_values(by="Target", inplace=True, ignore_index=True)
# Print the dataset
print(joined)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       102677  3.0  0.0    normal  abnormal  notpresent  notpresent  yes   
1       234299  4.0  0.0    normal  abnormal  notpresent  notpresent  yes   
2       636769  2.0  0.0    normal  abnormal     present     present   no   
3       484175  2.0  0.0  abnormal  abnormal     present  notpresent  yes   
4       481293  3.0  1.0    normal  abnormal     present  notpresent  yes   
..         ...  ...  ...       ...       ...         ...         ...  ...   
193     433613  0.0  0.0    normal    normal  notpresent  notpresent   no   
194     436056  0.0  0.0    normal    normal  notpresent  notpresent   no   
195     441170  0.0  0.0    normal    normal  notpresent  notpresent   no   
196     441170  0.0  0.0    normal    normal  notpresent  notpresent   no   
197     996076  0.0  0.0    normal    normal  notpresent  notpresent   no   

      dm cad  ...    bp    bgr     bu    sc    sod  pot  hemo   pcv     wbc

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [34]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
dummy = pd.get_dummies(joined, columns = categorical_columns, dtype=int)
# Print the dataset
print(dummy)

     unique_id   al   su   age    bp    bgr     bu    sc    sod  pot  ...  \
0       102677  3.0  0.0  59.0  70.0   76.0  186.0  15.0  135.0  7.6  ...   
1       234299  4.0  0.0  48.0  80.0   95.0  163.0   7.7  136.0  3.8  ...   
2       636769  2.0  0.0  50.0  90.0  128.0  208.0   9.2  134.0  4.8  ...   
3       484175  2.0  0.0  53.0  90.0   70.0  107.0   7.2  114.0  3.7  ...   
4       481293  3.0  1.0  60.0  60.0  288.0   36.0   1.7  130.0  3.0  ...   
..         ...  ...  ...   ...   ...    ...    ...   ...    ...  ...  ...   
193     433613  0.0  0.0  50.0  80.0   97.0   40.0   0.6  150.0  4.5  ...   
194     436056  0.0  0.0  12.0  80.0  100.0   26.0   0.6  137.0  4.4  ...   
195     441170  0.0  0.0  79.0  80.0  111.0   44.0   1.2  146.0  3.6  ...   
196     441170  0.0  0.0  79.0  80.0  111.0   44.0   1.2  146.0  3.6  ...   
197     996076  0.0  0.0  57.0  60.0  105.0   49.0   1.2  150.0  4.7  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_ye

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

It might work to just have a single column for the binary variable because the columns will represent parallel and exactly opposite vectors. However, when there's more than two options it won't work; in the sandwiches example, if we set ketchup=1, mustard = 2, and mayonnaise = 3, the model would assume that mustard is like mayochup, just an even mixture of ketchup and mayo.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [35]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
zscores = stats.zscore(dummy[numerical_columns])
notoutlier = zscores[abs(zscores) < 3].notnull().all(axis=1)
no_outliers = dummy[notoutlier]
no_outliers = no_outliers.reset_index(drop=True)
# Print the dataset
print(no_outliers)

     unique_id   al   su   age    bp    bgr     bu   sc    sod  pot  ...  \
0       234299  4.0  0.0  48.0  80.0   95.0  163.0  7.7  136.0  3.8  ...   
1       481293  3.0  1.0  60.0  60.0  288.0   36.0  1.7  130.0  3.0  ...   
2       474407  4.0  0.0  21.0  90.0  107.0   40.0  1.7  125.0  3.5  ...   
3       843362  4.0  0.0  71.0  60.0  118.0  125.0  5.3  136.0  4.9  ...   
4       828592  3.0  0.0  71.0  70.0  219.0   82.0  3.6  133.0  4.4  ...   
..         ...  ...  ...   ...   ...    ...    ...  ...    ...  ...  ...   
168     433613  0.0  0.0  50.0  80.0   97.0   40.0  0.6  150.0  4.5  ...   
169     436056  0.0  0.0  12.0  80.0  100.0   26.0  0.6  137.0  4.4  ...   
170     441170  0.0  0.0  79.0  80.0  111.0   44.0  1.2  146.0  3.6  ...   
171     441170  0.0  0.0  79.0  80.0  111.0   44.0  1.2  146.0  3.6  ...   
172     996076  0.0  0.0  57.0  60.0  105.0   49.0  1.2  150.0  4.7  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_yes  \
0      

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [36]:
# Normalize the all Numerical Attributes in the dataset.
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
no_outliers[numerical_columns] = no_outliers[numerical_columns].apply(lambda x: (x - x.min())/(x.max()-x.min()))
normalized = no_outliers
# Print the dataset
print(normalized)

     unique_id   al   su       age    bp       bgr        bu        sc  \
0       234299  4.0  0.0  0.545455  0.50  0.107296  1.000000  0.901235   
1       481293  3.0  1.0  0.701299  0.00  0.935622  0.169935  0.160494   
2       474407  4.0  0.0  0.194805  0.75  0.158798  0.196078  0.160494   
3       843362  4.0  0.0  0.844156  0.00  0.206009  0.751634  0.604938   
4       828592  3.0  0.0  0.844156  0.25  0.639485  0.470588  0.395062   
..         ...  ...  ...       ...   ...       ...       ...       ...   
168     433613  0.0  0.0  0.571429  0.50  0.115880  0.196078  0.024691   
169     436056  0.0  0.0  0.077922  0.50  0.128755  0.104575  0.024691   
170     441170  0.0  0.0  0.948052  0.50  0.175966  0.222222  0.098765   
171     441170  0.0  0.0  0.948052  0.50  0.175966  0.222222  0.098765   
172     996076  0.0  0.0  0.662338  0.00  0.150215  0.254902  0.098765   

          sod       pot  ...  cad_no  cad_yes  appet_good  appet_poor  pe_no  \
0    0.600000  0.310345  ...   

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: The unique_id column is arbitrary and unrelated to a patient's health, so it should be removed.

In [37]:
#Remove that columnnormalized
normalized = normalized.drop(columns = ["unique_id"])
# Print the dataset
print(normalized)

      al   su       age    bp       bgr        bu        sc       sod  \
0    4.0  0.0  0.545455  0.50  0.107296  1.000000  0.901235  0.600000   
1    3.0  1.0  0.701299  0.00  0.935622  0.169935  0.160494  0.428571   
2    4.0  0.0  0.194805  0.75  0.158798  0.196078  0.160494  0.285714   
3    4.0  0.0  0.844156  0.00  0.206009  0.751634  0.604938  0.600000   
4    3.0  0.0  0.844156  0.25  0.639485  0.470588  0.395062  0.514286   
..   ...  ...       ...   ...       ...       ...       ...       ...   
168  0.0  0.0  0.571429  0.50  0.115880  0.196078  0.024691  1.000000   
169  0.0  0.0  0.077922  0.50  0.128755  0.104575  0.024691  0.628571   
170  0.0  0.0  0.948052  0.50  0.175966  0.222222  0.098765  0.885714   
171  0.0  0.0  0.948052  0.50  0.175966  0.222222  0.098765  0.885714   
172  0.0  0.0  0.662338  0.00  0.150215  0.254902  0.098765  1.000000   

          pot      hemo  ...  cad_no  cad_yes  appet_good  appet_poor  pe_no  \
0    0.310345  0.207921  ...       1       

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [38]:
# Export the dataframe to a new csv file
normalized.to_csv("kidney_results.csv")
# Export the dataframe to a new json file
normalized.to_json("kidney_results.json")

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

### Paste the markdown table here```markdown
| | al | su | age | bp | bgr | bu | sc | sod | pot | hemo | pcv | wbcc | rbcc | rbc_abnormal | rbc_normal | pc_abnormal | pc_normal | pcc_notpresent | pcc_present | ba_notpresent | ba_present | htn_no | htn_yes | dm_no | dm_yes | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|----|------|------|---------|-------|---------|--------|--------|---------|---------|----------|---------|----------|----------|------------------|----------------|------------------|----------------|-------------------|----------------|------------------|----------------|------------|-------------|-----------|------------|------------|-------------|----------------|----------------|----------|------------|-----------|-------------|----------------|-----------------|
| 0 | 4.0 | 0.0 | 0.545455 | 0.50 | 0.107296 | 1.000000 | 0.901235 | 0.600000 | 0.310345 | 0.207921 | 0.290323 | 0.209677 | 0.114286 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | 1 |
| 1 | 3.0 | 1.0 | 0.701299 | 0.00 | 0.935622 | 0.169935 | 0.160495 | 0.428571 | 0.034483 | 0.019802 | 0.064516 | 0.879032 | 0.000000 | 0 | 1 | 1 | 0 | 0 | 1 | 1 | 0 | 0 | 1 | 1 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | 0 | 1 | 1 |
| 2 | 4.0 | 0.0 | 0.194805 | 0.75 | 0.158798 | 0.196078 | 0.160495 | 0.285714 | 0.206897 | 0.059406 | 0.000000 | 0.653226 | 0.257143 | 0 | 1 | 1 | 0 | 0 | 1 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 1 |
| 3 | 4.0 | 0.0 | 0.844156 | 0.00 | 0.206010 | 0.751634 | 0.604938 | 0.600000 | 0.689655 | 0.366337 | 0.387097 | 0.879032 | 0.371429 | 0 | 1 | 0 | 1 | 1 | 0 | 1 | 0 | 0 | 1 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 |
| 4 | 3.0 | 0.0 | 0.844156 | 0.25 | 0.639485 | 0.470588 | 0.395062 | 0.514286 | 0.517241 | 0.267327 | 0.322581 | 0.104839 | 0.171429 | 0 | 1 | 1 | 0 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 1 | 0 | 0 | 1 | 0 | 1 |
| 5 | 2.0 | 0.0 | 0.714286 | 0.50 | 0.442060 | 0.901961 | 0.432100 | 0.571429 | 0.793103 | 0.000000 | 0.032258 | 0.395161 | 0.057143 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 |
| 6 | 2.0 | 0.0 | 0.870130 | 0.50 | 0.785408 | 0.862745 | 0.518520 | 0.657143 | 1.000000 | 0.277228 | 0.322581 | 0.233871 | 0.371429 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 |
| 7 | 2.0 | 2.0 | 0.740260 | 1.00 | 0.901288 | 0.163399 | 0.345679 | 0.800000 | 0.206897 | 0.524752 | 0.548387 | 0.443548 | 0.342857 | 0 | 1 | 0 | 1 | 1 | 0 | 0 | 1 | 0 | 1 | 1 | 0 | 0 | 1 | 0 | 1 | 1 | 0 | 0 | 1 | 0 | 1 |
| 8 | 3.0 | 0.0 | 1.000000 | 0.25 | 0.137339 | 0.326797 | 0.271605 | 0.000000 | 0.965517 | 0.099010 | 0.096774 | 0.685484 | 0.028571 | 0 | 1 | 0 | 1 | 1 | 0 | 1 | 0 | 0 | 1 | 1 | 0 | 1 | 0 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 |
| 9 | 4.0 | 3.0 | 0.818182 | 0.25 | 0.618026 | 0.562092 | 0.728395 | 0.142857 | 0.344828 | 0.168317 | 0.161290 | 0.580645 | 0.085714 | 0 | 1 | 1 | 0 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 |
| 10 | 4.0 | 2.0 | 0.753247 | 1.00 | 0.399142 | 0.287582 | 0.839506 | 0.714286 | 0.586207 | 0.019802 | 0.096774 | 0.258065 | 0.114286 | 1 | 0 | 1 | 0 | 1 | 0 | 0 |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

Honestly, by not using an LLM for this (383GPT got stuck before it got through the full 15 rows). But more seriously, maybe by specifically telling it to copy the data exactly as it was given? But given how hard it is to check manually I think this would be done better using the pandas .to_markdown method.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [39]:
# Code to rename all the columns in the dataset

column_mapping = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    # 'pcc': 'pus cell clumpsssdddfsdfa
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}

# Rename the columns ung the mapping
# dfsdnormalized.rename(columns=column_mapping, inplace=True)
print(normalized)

      al   su       age    bp       bgr        bu        sc       sod  \
0    4.0  0.0  0.545455  0.50  0.107296  1.000000  0.901235  0.600000   
1    3.0  1.0  0.701299  0.00  0.935622  0.169935  0.160494  0.428571   
2    4.0  0.0  0.194805  0.75  0.158798  0.196078  0.160494  0.285714   
3    4.0  0.0  0.844156  0.00  0.206009  0.751634  0.604938  0.600000   
4    3.0  0.0  0.844156  0.25  0.639485  0.470588  0.395062  0.514286   
..   ...  ...       ...   ...       ...       ...       ...       ...   
168  0.0  0.0  0.571429  0.50  0.115880  0.196078  0.024691  1.000000   
169  0.0  0.0  0.077922  0.50  0.128755  0.104575  0.024691  0.628571   
170  0.0  0.0  0.948052  0.50  0.175966  0.222222  0.098765  0.885714   
171  0.0  0.0  0.948052  0.50  0.175966  0.222222  0.098765  0.885714   
172  0.0  0.0  0.662338  0.00  0.150215  0.254902  0.098765  1.000000   

          pot      hemo  ...  cad_no  cad_yes  appet_good  appet_poor  pe_no  \
0    0.310345  0.207921  ...       1       

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [40]:
# Converted SQL to Pandas code
# # Sample DataFrame creation (replace this with your actual DataFrame)
# data = {
#     'Target': ['ckd', 'notckd', 'sckd', 'notckd', 'ckd', 'notckd'],
#     # other columns can be added here as needed
# }

# df = pd.DataFrame(data)

# # Grouping by 'Target' and counting the occurrences
result = normalized.groupby('Target_ckd').size().reset_index(name='count')

# Display the result
print(result)

   Target_ckd  count
0           0    150
1           1     23


*Describe what the above code does here*
I had to make some modifications to the code to use the correct database and because it tried to turn column headings (and some strings that are just close to column headings, like sckd) into data points. All in all,
 this code bundles together all the rows with matching target_ckd values, finds the size of each group (aka the number of data points representing that categorization of having CKD or not), and converts it back into a dataframe in which the left column represents the category (CKD or not) and the right column represents the number of occurrences.